### ISSUE TITLE
Neural network internal data type

### ISSUE THREAD

**Issue by raver119** [eclipse/deeplearning4j#6866]
As discussed with @AlexDBlack in pm:

Since now we support lots of different data types, we should provide an option to configure data type for NN params/activations via builder options. At least at global NN level.

Aha! Link: https://skymindai.aha.io/features/DL4J-39

**Comment by AlexDBlack** [comment]
Some design questions and considerations:
1. Do we allow separate configuration of activations and parameters datatypes - i.e., FP32 params, FP16 activations/activation gradients
2. Do we allow setting on a per-layer level? (uncommon in practice, but it probably makes sense - example: FP16 embedding layer, FP32 most layers, FP64 loss layer)
3. Do we provide utilities or tools for conversion - i.e., convert trained FP32 to FP16 - or some sort of "load as FP16" option for model deserialization?

First step is probably simple single (fixed) global network datatype setting, which controls datatype of parameters, activations and gradients etc.
